### ISSUE TITLE
Buffered I/O Abstraction

### ISSUE THREAD

**Issue by lu-zero** [rust-av/rust-av#13]
The Rust BufRead implementation does not provide a `peek()` method.

To support sane probing it would be quite useful.

**Comment by lu-zero** [comment]
@tmatth so far I'm just using BufRead with a buffer size that "is enough to probe" would be nice having a real peek() implementation maybe...

**Comment by lu-zero** [comment]
Alternatively the demuxing context could act as a specialized BufRead-like structure.

**Comment by lu-zero** [comment]
Now there is an extra simple and stupid trait to have some byte-specific `peek()` functions for the joy of @slomo

https://github.com/rust-av/rust-av/commit/08dce83291cba397b3fb4cb4af0b96bb0910bfaa

**Comment by FreezyLemon** [comment]
It looks like `BufRead::fill_buf` is pretty much a `peek` [equivalent](https://doc.rust-lang.org/std/io/trait.BufRead.html#tymethod.fill_buf):

> When calling this method, none of the contents will be “read” in the sense that later calling read may return the same contents.

Example from the docs:
```rust
let buffer = stdin.fill_buf().unwrap();

// work with buffer
println!("{buffer:?}");

// ensure the bytes we worked with aren't returned again later
let length = buffer.len();
stdin.consume(length);
```

**Comment by Luni-4** [comment]
Good point! We can use `fill_buf` then creating some tests to verify that works
